File: 1268-search-suggestions-system/1268-search-suggestions-system.swift
class Solution {
    let rootNode = Node(root: "-")
    
    func getWords(_ curr: Node) -> [[Character]] {
        var ret = [[Character]]()
        var q = [(curr, [Character]())]
        while !q.isEmpty {
            let (curr, wordArr) = q.removeFirst()
            if curr.wordEnd {
                ret.append(wordArr)
            }
            for c in curr.children.values {
                q.append((c, wordArr + [c.root]))
            }
        }
        return ret
    }

    func suggestedProducts(_ products: [String], _ searchWord: String) -> [[String]] {
        for p in products { insertTrie(p) }
        var ret = [[String]]()
        var curr = rootNode
        var prefix = [Character]()
        for (i, c) in Array(searchWord).enumerated() {
            var temp = [String]()
            prefix.append(c)
            if let found = curr.children[c] {
                curr = found
                for suffix in getWords(curr) {
                    let builtWord = String(prefix + suffix)
                    temp.append(builtWord)
                }
                temp.sort()
                temp = Array(temp[..<min(temp.count, 3)])
                ret.append(temp)
            } else {
                for j in i ..< searchWord.count {
                    ret.append([])
                }
                break
            }
        }
        return ret
    }
    
    func insertTrie(_ p: String) {
        var curr = rootNode
        for el in p {
            curr = curr.children[el] ?? {
                curr.children[el] = Node(root: el)
                return curr.children[el]!
            }()
        }
        curr.wordEnd = true
    }
}

class Node: CustomStringConvertible {
    var root: Character
    var children: [Character: Node] = [:]
    var wordEnd: Bool = false

    init(root: Character) {
        self.root = root
    }

    var description: String {
        return "[\(root): \(children)]"
    }
}

File: 1268-search-suggestions-system/1268-search-suggestions-system.py
class Solution:
    def suggestedProducts(self, products, searchWord):
        for p in products:
            self.insertTrie(p)
        ret = []
        curr = self.rootNode
        prefix = []
        for (i, char) in enumerate(searchWord):
            prefix.append(char)
            temp = []
            if curr.children[self.x(char)]:
                curr = curr.children[self.x(char)]
                for suffix in self.getWords(curr):
                    builtWord = "".join(prefix + suffix)
                    temp.append(builtWord)
            else:
                while i < len(searchWord):
                    ret.append([])
                    i += 1
                break
            ret.append(temp)

        return ret

    def __init__(self):
        self.rootNode = Node("-")

    def getWords(self, curr):
        def getWordsBacktrack(curr, built):
            if len(ret) >= 3:
                return
            if curr.wordEnd:
                ret.append(built)
            for c in curr.children:
                if c: getWordsBacktrack(c, built + [c.root])
        ret = []
        getWordsBacktrack(curr, [])
        return ret

    def insertTrie(self, p):
        curr = self.rootNode
        for c in p:
            if not curr.children[self.x(c)]:
                curr.children[self.x(c)] = Node(c)
            curr = curr.children[self.x(c)]
        curr.wordEnd = True
        
    def x(self, ch):
        return ord(ch) - ord('a')

class Node:
    def __init__(self, root):
        self.root = root
        self.children = [None] * 26
        self.wordEnd = False

    def __repr__(self):
        return f"[{self.root}: {[x for x in self.children if x is not None]}]"
